### ISSUE TITLE
"Uncaught TypeError: root is undefined" when importing pdfjsLib from 'pdfjs-dist/webpack'

### ISSUE THREAD

**Issue by RST-J** [mozilla/pdf.js#12786]
I am a bit heavily lost in the quest for a solution to my problem. After getting nowhere by trying all sorts of stuff I thought might help for some days, I started my quest for an answer or hint with an [issue in the webpacker repo](https://github.com/rails/webpacker/issues/2804). They said it does not look like a particular issue with webpacker and sent me here. So I hope, someone here can help me out.

I want to setup pdf.js in our Rails app and followed the [recommendation for webpack users](https://github.com/mozilla/pdf.js/blob/v2.6.347/examples/webpack/README.md) which basically says to simply use `import pdfjsLib from 'pdfjs-dist/webpack';` which takes care of setting up the worker. With this code in my JavaScript I get the following error in my browser console when I access a page that triggers this code:
```
Uncaught TypeError: root is undefined
    webpackUniversalModuleDefinition pdf.js:77
    js                               pdf.js:78
```
I cannot find the respective code in the pdfjs-dist repository, I assume it is generated by some build/release pipeline. But here is what the respective file in `node_modules/pdfjs-dist/build/pdf.js` looks like. It starts with something that looks webpack specific:
```javascript
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define("pdfjs-dist/build/pdf", [], factory);
	else if(typeof exports === 'object')
		exports["pdfjs-dist/build/pdf"] = factory();
	else
		root["pdfjs-dist/build/pdf"] = root.pdfjsLib = factory();
})(this, function() {
// all the code
});
```
The flagged lines belong to the body of a function defined like this:
```javascript
// define __esModule on exports
__w_pdfjs_require__.r = function(exports) {
  if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {                  // <-- Line 77
    Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' }); // <-- Line 78
  }
  Object.defineProperty(exports, '__esModule', { value: true });
};
```
A gist with the complete pdf.js from the pdfjs-dist package can be viewed [here](https://gist.github.com/RST-J/109f611834b95a5699f9c41d5f7e6cdc)
**Note:** The gist refers to version 2.5.207. In the meantime I also tried 2.6.347 with the exact same result (same error message in the same line with the same code).

That is my situation and I am lost between webpacker, webpack and pdf.js. My intuition was that I probably need to tell webpack(er?) that it should inject something for the `root` value. However, this is a formal parameter of the module definition as far as I see it. So what value for what keyword would I have to inject exactly and where?
I have worker-loader installed, that part seems to work (got an error related to that before).

I think/hope/assume that it should be possible to somehow only do what the README for pdf.js says: Import the thing via `webpack.js`. After all it is advertised as `the zero-configuration method for Webpack users` and I would like to be one of that happy users. Maybe I am only missing something very simple that is painfully obvious if you know it ... but I just don't know and couldn't figure out in three days so far. So, anyone a hint or maybe even the solution? :smiley: 

Attach (recommended) or Link to PDF file here:

Configuration:
- Web browser and its version: FF 84.0
- Operating system and its version: Kubuntu 20.04 (Rails server runs in docker container though)
- PDF.js version: 2.5.207, 2.6.347
- Is a browser extension: No

Steps to reproduce the problem:
1. Not applicable yet. I could setup a dummy Rails app showing the problem, if that would be required. But that seems a bit much to me and if that is actually the problem then it maybe is not a pdfjs or pdfjs-only issue(?).

**Comment by timvandermeij** [comment]
Did you try the Webpack example standalone, i.e., outside of Webpacker/Rails? If that works, and it should, then it's related to the integration with Webpacker/Rails and that's unfortunately something we don't know anything about and can't provide meaningful assistance for.

**Comment by Genkilabs** [comment]
WORKAROUND for running pdf.js on Rails 6.1 Ruby 3.0 with webpacker. 
This Gist is not as good as if pdf.js exported correctly, but it will get you through the day.

https://gist.github.com/Genkilabs/a5285020ca548f589b9c46497afca941

NOTE: I did NOT do anything like `require('pdfjs-dist/es5/build/pdf')` in my application.js and I did not `import` it in my module, because its a globally-scoped kludgearound!
